### ISSUE TITLE
Registration using 3+ disc labels results in distorted top portion of warped template

### ISSUE THREAD

**Issue by yw7** [spinalcordtoolbox/spinalcordtoolbox#4085]
Hello SCT Team,

I am encountering an issue with the `sct_register_to_template` command. When I use the default parameters to register an image to the PAM50 template, I have noticed that parts of the resulting registered image extend beyond the template boundaries. Specifically, the vertebrae T5 label (12) appears to be extending into the medulla.

I am currently using SCT version 5.8 on Windows (native).

![image](https://user-images.githubusercontent.com/36595323/230679517-63fb280b-9512-4aa8-91fd-36b7fd9b93c5.png)

Here’s the command I used:
```
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg.nii.gz -ldisc sub-errsm37_T1w_labels-disc-manual.nii.gz -c t1 -qc qc
sct_warp_template -d sub-errsm37_T1w.nii.gz -w warp_template2anat.nii.gz -a 0 -s 1 -qc qc
```

`errsm37` is the sample subject for the [`totalsegmentator-mri`](https://github.com/neuropoly/totalsegmentator-mri) project, and the images can be downloaded from [this Google Drive link](https://drive.google.com/drive/folders/12PhGm1C2vpeenfL79HH9ZYk-7O3N_uBr).

I’m wondering if I might be missing something in the registration process. Could you please provide any guidance or suggestions for resolving this issue? I will send the relevant files via google drive on Slack.

As always, I greatly appreciate your help and support.

Best regards,

Yehuda

Forum post [here](https://forum.spinalcordmri.org/t/issue-with-sct-register-to-template-template-extending-template-boundaries-on-image/1056).

**Comment by valosekj** [comment]
Hi @yw7, thank you for reaching out!

I am able to reproduce the behaviour with the master SCT version on macOS.

The issue is coming from the `sct_register_to_template` command. The warped PAM50 (`template2anat.nii.gz` file) produced by the `sct_register_to_template` contains a weird part above the C1 level; see the red circle in the image below.

<img width="1408" alt="Screenshot 2023-04-08 at 10 20 46 AM" src="https://user-images.githubusercontent.com/39456460/230711650-ac1265d5-7d5f-4fec-8d1b-3ffbad20d8f8.png">

We will continue with the investigation.

**Comment by yw7** [comment]
Thank you @valosekj for the quick replay and for the investigation,

I also noticed that the error in the registration exists not only at the C1 level but also at C7-T1 and T12 levels, as shown in the attached picture. It would be very helpful if you could investigate these issues as well.

![image](https://user-images.githubusercontent.com/36595323/230728473-411fa3db-a902-4b62-9cb1-05d438d768d2.png)

Thank you very much,
Yehuda

**Comment by yw7** [comment]
Thank you @valosekj for the quick replay and for the investigation,

I also noticed that the error in the registration exists not only at the C1 level but also at C7-T1 and T12 levels, as shown in the attached picture. It would be very helpful if you could investigate these issues as well.

![image](https://user-images.githubusercontent.com/36595323/230728598-8b370398-b02f-4d8d-8e39-9fb88eb7dea0.png)

Thank you very much,
Yehuda

**Comment by jcohenadad** [comment]
@yw7 the syntax you wrote here did not correspond to the images you sent (`sub-errsm37_T1w_seg_labeled_discs_manual.nii.gz` does not exist-- I presume you referred to `sub-errsm37_T1w_labels-disc-manual.nii.gz`), so I edited your post to avoid confusion

**Comment by jcohenadad** [comment]
Checked label values-- seems OK:
```console
julien-macbook:~/temp/20230409_sct-bug-yehuda $ sct_label_utils -i sub-errsm37_T1w_labels-disc-manual.nii.gz -display

--
Spinal Cord Toolbox (git-master-7ead83200d7ad9ee5e0d112e77a1d7b894add738)

sct_label_utils -i sub-errsm37_T1w_labels-disc-manual.nii.gz -display
--

Position=(87,122,562) -- Value= 1
Position=(87,116,545) -- Value= 2
Position=(87,114,532) -- Value= 3
Position=(87,112,516) -- Value= 4
Position=(87,112,501) -- Value= 5
Position=(87,111,486) -- Value= 6
Position=(87,110,472) -- Value= 7
Position=(87,108,457) -- Value= 8
Position=(87,104,443) -- Value= 9
Position=(87,96,426) -- Value= 10
Position=(87,88,409) -- Value= 11
Position=(90,84,384) -- Value= 12
Position=(90,78,365) -- Value= 13
Position=(90,76,341) -- Value= 14
Position=(90,75,317) -- Value= 15
Position=(90,76,293) -- Value= 16
Position=(90,80,268) -- Value= 17
Position=(90,83,243) -- Value= 18
Position=(90,88,217) -- Value= 19
Position=(90,94,188) -- Value= 20
Position=(90,101,158) -- Value= 21
```

**Comment by jcohenadad** [comment]
I can also reproduce the issue when looking at the registered PAM50 in the subject's space. However, the registration in the PAM50 space looks fine:

![anim](https://user-images.githubusercontent.com/2482071/230784155-994d6833-e69b-4c38-927d-1db6a7293cc7.gif)

**Comment by jcohenadad** [comment]
The issue is very likely caused by the use of more than 2 labels, which implies non-linear alignment of discs without extrapolation outside the top and bottom labels. Related to: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2281

**Comment by yw7** [comment]
Hello @jcohenadad,
Thank you very much for investigating this issue; is there a way to crop the result `warp_template2anat.nii.gz` in the z-axis between the top and bottom labels?

**Comment by jcohenadad** [comment]
Warping fields after decomposing into X/Y/Z:
```console
sct_image -i warp_template2anat.nii.gz -mcs
```

Large discontinuity for the Z component:
![image](https://user-images.githubusercontent.com/2482071/230785695-4c931b11-203a-4d0b-85af-3d19c35b5f4e.png)

Suspicious values outside the masked spinal cord (10^38):
![image](https://user-images.githubusercontent.com/2482071/230785731-d72906a0-cab3-497a-9133-f6d09de0bca7.png)

**Comment by jcohenadad** [comment]
The segmentation is highly discontinuous below a certain point, and given that the straightening algorithm will consider the top/bottom label of the segmentation as the edges of the segmentation that is used to calculate the straightening field, these discontinuities will create unreliable straightening:

![Screen Shot 2023-04-09 at 12 59 03 PM](https://user-images.githubusercontent.com/2482071/230786108-23e6551a-b0e5-44d8-9639-ffedbe0a98ca.png)

**Comment by jcohenadad** [comment]
Trying to crop the segmentation below z=187 (where discontinuities start to appear), and re-run the registration to the template:
```console
sct_crop_image -i sub-errsm37_T1w_seg.nii.gz -zmin 187 -o sub-errsm37_T1w_seg_cropz.nii.gz 
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg_cropz.nii.gz -ldisc sub-errsm37_T1w_labels-disc-manual.nii.gz -c t1 -qc qc -r 0 -v 2
```

**Comment by jcohenadad** [comment]
old reg:
![old](https://user-images.githubusercontent.com/2482071/230787079-b9ccfdde-16e9-4628-a29e-19261845f2d5.gif)

new reg:
![new](https://user-images.githubusercontent.com/2482071/230787078-27852b76-58b3-4cc5-b516-cfff303f7d11.gif)

solved the "bottom" part, but weird things happening at the top. investigating...

**Comment by jcohenadad** [comment]
the labels (in red) are quite unusual because (i) they are located in the posterior tip of the disc (vs. in the spinal cord) and (ii) there is a label at the top of C1. This scenario is poorly tested, so I suspect something is incompatible with our registration pipeline. This needs to be thoroughly debugged, notably:
- what happens to these labels during straightening
- what happens to the label-based registration (step0)
- how problematic is the padding added on the top/bottom
- how is the final warping field generated, and how the 'unseen' regions are filled up in the warping field.

**Comment by joshuacwnewton** [comment]
> the labels (in red) are quite unusual because (i) they are located in the posterior tip of the disc (vs. in the spinal cord) and (ii) there is a label at the top of C1.

Also, (iii) the labels are all located to the left of the spinal cord centerline, in addition to being at the posterior tip: 

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/34f5b228-19cd-4c92-ab68-35d26553cf39)

**Comment by joshuacwnewton** [comment]
Small note: I am trying to re-run the registration command using the images provided via @yw7's [Google Drive](https://drive.google.com/drive/folders/12PhGm1C2vpeenfL79HH9ZYk-7O3N_uBr) link:

```
sct_register_to_template -i sub-errsm37_T1w.nii.gz \
                         -s sub-errsm37_T1w_seg.nii.gz \
                         -ldisc sub-errsm37_T1w_labels-disc-manual.nii.gz \ 
                         -c t1 -qc qc -r 0 -v 2
```

However, I am running into the following error, occurring in the `-v 2` plotting code:

```
  File "/home/joshua/repos/spinalcordtoolbox/spinalcordtoolbox/registration/algorithms.py", line 866, in register2d_centermassrot
    coord_dest_rot = coord_dest[iz] * R.T
ValueError: operands could not be broadcast together with shapes (51,2) (2,2) 
```

This error occurs regardless of the segmentation crop. I'm not sure why I've run into this when @jcohenadad has not? :thinking: 

For now I can just remove `-v 2` and debug normally.

**Comment by jcohenadad** [comment]
very interesting, thank you for the digging @joshuacwnewton !

**Comment by joshuacwnewton** [comment]
tl;dr: The distortions at the top of the segmentation are caused by using >2 labels, and NOT because of the presence of the extra C1 label or the position of the labels. In fact, `sct_register_to_template` handles these labels quite well when we limit it to only 2 labels!

----

Here are some tests I performed to come to this conclusion:

<details>
<summary> All labels (1.0-21.0) vs. All labels (No label 1.0) vs. All Labels (No labels 1.0/2.0) </summary>

Commands:

```bash
# All labels (1.0-21.0)
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg.nii.gz -ldisc sub-errsm37_T1w_labels-disc-manual.nii.gz -c t1 -qc qc -r 0 -ofolder no-crop_all-labels

# All labels (No 1.0)
sct_label_utils -i sub-errsm37_T1w_labels-disc-manual.nii.gz -remove 1 -o sub-errsm37_T1w_labels-disc-manual-remove-1.nii.gz 
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg_cropz.nii.gz -ldisc sub-errsm37_T1w_labels-disc-manual-remove-1.nii.gz -c t1 -qc qc -r 0 -ofolder cropped_remove-1

# All labels (No 1.0/2.0)
sct_label_utils -i sub-errsm37_T1w_labels-disc-manual.nii.gz -remove 1,2 -o sub-errsm37_T1w_labels-disc-manua-remove-1-2l.nii.gz 
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg_cropz.nii.gz -ldisc sub-errsm37_T1w_labels-disc-manual-remove-1-2.nii.gz -c t1 -qc qc -r 0 -ofolder cropped_remove-1-2
```

All labels (1.0-21.0):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/40ba95c7-c551-4e9f-b06b-57ddcf8d4e84)

All labels (No 1.0):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/a9f68f7f-c9de-488c-b463-e234920c0a11)

All labels (No 1.0/2.0):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/b97b8c25-3697-45c2-87ab-e594bd344b24)

As you can see, removing the C1/C2 labels does not eliminate the distorted top portion -- it simply shifts it downward.

</details>

<details>
<summary> All labels vs. 2 labels (1, 11) </summary>

Commands:

```bash
# 2 labels (1, 11)
sct_label_utils -i sub-errsm37_T1w_labels-disc-manual.nii.gz -keep 1,11 -o sub-errsm37_T1w_labels-disc-manual-keep-1-11.nii.gz 
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg_cropz.nii.gz -ldisc sub-errsm37_T1w_labels-disc-manual-keep-1-11.nii.gz -c t1 -qc qc -r 0 -ofolder cropped_keep-1-11
```

2 labels (1, 11):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/04065adc-4ee8-4a79-80bc-01f04047bf74)

Even though we are using the first label (1.0) as one of our landmarks, we are still able to eliminate the distortion in the warped template by sticking to only 2 labels.

</details>

<details>
<summary> All labels (manual disc, 1.0-21.0) vs. All labels (vert body, 2.0-19.0) </summary>

Commands:

```bash
# Generate labeled segmentation
sct_label_vertebrae -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg_cropz.nii.gz -c t1 -ofolder labeled-vertebrae
# Extract all vert-body labels (1-19, excluding 20 because the verterbral label is only partial)
sct_label_utils -i labeled-vertebrae/sub-errsm37_T1w_seg_cropz_labeled.nii.gz -vert-body 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19 -o sub-errsm37_T1w_labels-vertbody-automatic.nii.gz
# Remove label 1 because of bug (https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4139)
sct_label_utils -i sub-errsm37_T1w_labels-vertbody-automatic.nii.gz -remove 1 -o sub-errsm37_T1w_labels-vertbody-automatic-remove-1.nii.gz
# Registration (note the use of `-lspinal` instead of `-ldisc`)
sct_register_to_template -i sub-errsm37_T1w.nii.gz -s sub-errsm37_T1w_seg_cropz.nii.gz -lspinal sub-errsm37_T1w_labels-vertbody-automatic-remove-1.nii.gz -c t1 -qc qc -r 0 -ofolder cropped_all-labels_vertbody_remove-1
```

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/ba688082-6d3a-4162-80cc-b1f5fdaa852b)

There is no significant improvement/difference to the top of the warped template, even when vert body labels are used. The performance is pretty much identical to the "No 1.0" disc label image from the first test:

> ![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/a9f68f7f-c9de-488c-b463-e234920c0a11)

</details>

----

Currently, [in our tutorials](https://spinalcordtoolbox.com/user_section/tutorials/registration-to-template/vertebral-labeling/how-many-labels-for-registration.html), we say the following:

> * 3+ Labels: This is the most accurate method for matching the vertebral levels of the anatomical image to the levels of the template. However, there is a key downside to this approach: Because more than two points must be matched, the level-matching transformation cannot be affine. As a result, the output warping field will be undefined for regions above the top label and below the bottom label.

Instead of "undefined", however, we get this strange stretched/warped portion. So, if necessary, I can explore further why using >2 labels causes this issue. :thinking:

**Comment by yw7** [comment]
Hello SCT Team,

Just wondering if there's a quick fix or workaround I could use to get rid of that extra part at the top until there will be an official solution?

Thanks a lot !!
yehuda

**Comment by jcohenadad** [comment]
Sorry for dropping the ball on this @yw7. Yes, we will go back to this issue ASAP. 

@joshuacwnewton I am wondering if the updated PAM50 (https://github.com/spinalcordtoolbox/PAM50/pull/21) can solve it?

**Comment by joshuacwnewton** [comment]
> Just wondering if there's a quick fix or workaround I could use to get rid of that extra part at the top until there will be an official solution?

It's a bit buried in my comment [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4085#issuecomment-1594906538), but the short-term solution I found was to only use 2 labels during registration. _(If you click on the drop-down called "All labels vs. 2 labels (1, 11)", it should show an example command with results -- no buggy top part.)_

((Also, if you happen to perform cropping, then `-b 0` is necessary as per my comment [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4085#issuecomment-1593519336).))

Please let me know if the registration results using 2 labels are not good enough for your needs. In that case, I can start looking into preventing the issue from happening when using 3+ labels.

**Comment by joshuacwnewton** [comment]
Small update to @jcohenadad's earlier comment re: Z discontinuities here: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4085#issuecomment-1501168961

`warp_template2anat.nii.gz` is composed of [both `warp_straight2curve.nii.gz` and the `warp_inverse` file](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/scripts/sct_register_to_template.py#L734) (template -> straightened anat). 

So, I decided to look at both warping fields individually, and... the registration warping field (right) appears to be fine? The issue instead appears to stem from the straight2curve warping field (Z component displayed left).

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/ae8cfb5a-2394-4596-a733-48733bdb2208)

This is kind of intuitive, right? Because the registration warping field (right) is slicewise in this case, the Z component is 0 (double-checked using `-mcs`). So any [discontinuities in the Z component](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4085#issuecomment-1501168961) must be due to the straightening step, I think?

This (I think) explains why the fixes to the PAM50 template have nothing to do with this -- the problem lies within the straightening instead. I will try to dig further into that part of the process. :)

**Comment by joshuacwnewton** [comment]
Small update to @jcohenadad's https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4085#issuecomment-1501168961 about Z discontinuities in `warp_template2anat.nii.gz`.

`warp_template2anat.nii.gz` is composed using 2 separate warping fields ("template -> straight anat" + "straight2curve" ):

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f5de501d3fe98c53861b54c60b9f5f5b96732772/spinalcordtoolbox/scripts/sct_register_to_template.py#L734

So, I decided to look at both of these warping fields individually, comparing the files from the successful registration (2 labels) with the faulty registration (3+ labels).

First, I decided to compare just the the registration warping fields (`warp_inverse` i.e. "template -> straight anat"):

- In both cases, the warping fields are slicewise transformations with Z component = 0.
- The main difference between the two is that the "3+ labels" warping field is smaller by 13 axial slices (z = 221 vs z = 234)
- However, the Z translation component of the affine matrix compensates for this:
   - "2 labels":  `-534.840027mm`
   - "3+ labels":  `-508.840027mm`
   - Difference: `+26mm` (which is equal to 13vox * 2mm/vox Z resolution)
- Apart from this, the affine matrices are identical, meaning that the registration warping fields should be interchangeable.
- So, given that there is no real effect in the Z axis, yet we see Z discontinuities in the final `warp_template2anat.nii.gz`, my hunch is that the registration warping field is not the culprit here.
- To test this hunch, I combined the "good" registration warping field (2 labels) with the "bad" straightening (3+ labels), and we still see the discontinuity.
- This (I think) explains why the [fixes to the PAM50 template](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4085#issuecomment-1671479078) have nothing to do with this.

So, I think there are 2 remaining culprits:

- The straightening warping field
- The call to `isct_ComposeMultiTransform` itself (due to excluded `-i straight2templateAffine.txt` argument)

**Comment by jcohenadad** [comment]
These are excellent investigations, @joshuacwnewton, I would not have done a better job. I agree that the curve2straight is suspicious, and could be the cause for this issue. I am surprised however that we don't encounter this issue more often. But maybe this is simply because most users work with 2 labels only. I think what _might_ be happening, as you also suspected, is that the size of the warping field (or the presence of zeroes in the parts where the bug occurs) could be the cause of the problem. I'm not sure how to address that though, but more thorough investigations are definitely needed. I'm also tagging @benjamindeleener who I'm sure encountered this issue in the past.

**Comment by joshuacwnewton** [comment]
I've found the problem! :smile: 

My first observation was that the top portion of the warped image looked to be a coherent representation of the spinal cord, just at some other Z location, seemingly lower down (due to the darker intensity). So, my hunch was that something in the straightening code was causing lower slices to be placed at the very top of the warped image. 

<details>

![1](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/5d0cb90f-04ac-4577-b1c3-0dadde57f646)

</details>

I then traced the code, looking at intermediate variables for anything that might represent this behavior, when I ended up at the [curve2straight lookup table](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d6aeaef52466129b5f24a56fa250403839bd0166/spinalcordtoolbox/straightening.py#L381-L391) code.

The curved (cropped) cord image has 382 slices, and the straightened image has 951 slices, so `lookup_curve2straight` is a mapping from 382 indices to 951 possible slices. For example:

> ![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/8971c108-5fe8-496c-943f-0b1ff21306f9)
> ![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/914a5394-d5e9-4530-a7b0-c291bd27c175)

However, curiously, the last 5 indices map to... themselves? 

> ![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/c9a208ed-1f64-4502-b14d-fb443b3f1c15)

This looked to me like a match to me! Since, I _think_ the last 5 curved slices in this case would be mapped to the lower-middle region of the straight space (~380 out of 952).

To test, then, I manually set these last 5 entries in the lookup table to 0. And, the issue is fixed! :tada:

![ezgif-1-847466b6ea](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/c1549209-ee0b-4e78-a94e-9a6a321b3c25)

The reason these lookup table entries were nonzero in the first place is due to this code:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d6aeaef52466129b5f24a56fa250403839bd0166/spinalcordtoolbox/types.py#L484-L487

I want to better understand what this code is actually doing, and why it's returning a slice index that's very far from the expected region. It seems like it was recently cleaned up in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4075, so perhaps that might have something to do with it? :thinking:

**Comment by jcohenadad** [comment]
These are AWESOME investigations @joshuacwnewton ! Thank you so much for doing this!!! tagging @benjamindeleener who I'm sure will also appreciate

**Comment by joshuacwnewton** [comment]
:tada: 

I plan on submitting a fix for this ASAP (it's an action item of mine from last week's SCT meeting).

**Comment by joshuacwnewton** [comment]
> https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d6aeaef52466129b5f24a56fa250403839bd0166/spinalcordtoolbox/types.py#L484-L487
>
> I want to better understand what this code is actually doing, and why it's returning a slice index that's very far from the expected region.

I spent quite a bit of time refactoring this code and adding extensive comments to make it easier for myself (and others) to understand. However, after all that refactoring, it turns out that the actual cause of the issue was _incredibly_ simple:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/33c924edc56d3e2f67e50b281c6e9b302afedf18/spinalcordtoolbox/straightening.py#L388-L389

- For `backup_index`, we pass `index`, which is from the curved centerline. 
- For `backup_centerline`, however, we pass the straight centerline. 
- This means that the logic in `get_closest_from_absolute_index` is combining two different spaces entirely.
- And, we can fix this by passing the curved `centerline` to the function instead. 
 
Doing so produces a lookup table that matches expectations (monotonically increasing slice values in the straightened space):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/2e32aa64-9f84-4a4d-9d73-46ab50a089c5)

And the registration results look great (contrast increased to show vertebral matching):

![ezgif com-gif-maker](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/8249ee2c-d884-4869-8610-1ecfbfeb6625)

This is a much more thorough solution than my quick test of "replace with all zeros". :stuck_out_tongue_closed_eyes: 

----

Interestingly enough, this bug was also the true cause of https://github.com/neuropoly/template/issues/60, meaning that the fix from https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4192 is wholly unnecessary and can be reverted, too.

**Comment by joshuacwnewton** [comment]
@yw7 Could you please try out the branch from PR #4206? ([`jn/4085-fix-distorted-registration-3-labels`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/jn/4085-fix-distorted-registration-3-labels))

**Comment by yw7** [comment]
> @yw7 Could you please try out the branch from PR #4206? ([`jn/4085-fix-distorted-registration-3-labels`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/jn/4085-fix-distorted-registration-3-labels))

Thank you so much @joshuacwnewton !!
It works perfect!
